File: 1411.number-of-ways-to-paint-n-3-grid.go
/*
 * @lc app=leetcode id=1411 lang=golang
 *
 * [1411] Number of Ways to Paint N × 3 Grid
 */
package main

// @lc code=start
func numOfWays(n int) int {
	ret1, ret2 := int64(6), int64(6)
	var mod int64 = 1e9 + 7

	for i := 0; i < n-1; i++ {
		ret1, ret2 = ((2*(ret1%mod)%mod)+(2*(ret2%mod)%mod))%mod, ((2*(ret1%mod)%mod)+(3*(ret2%mod)%mod))%mod
	}

	return int(((ret1 % mod) + (ret2 % mod)) % mod)
}

// @lc code=end

File: 1411.number-of-ways-to-paint-n-3-grid.rs
/*
 * @lc app=leetcode id=1411 lang=rust
 *
 * [1411] Number of Ways to Paint N × 3 Grid
 */

// @lc code=start
impl Solution {
    pub fn num_of_ways(n: i32) -> i32 {
        let mut ret1: i64 = 6;
        let mut ret2: i64 = 6;
        let m: i64 = 1_000_000_007;

        for _ in 0..n - 1 {
            let tmp1 = ((2 * (ret1 % m) % m) + (2 * (ret2 % m) % m)) % m;
            let tmp2 = ((2 * (ret1 % m) % m) + (3 * (ret2 % m) % m)) % m;

            ret1 = tmp1;
            ret2 = tmp2;
        }

        (((ret1 % m) + (ret2 % m)) % m) as i32
    }
}
// @lc code=end
